src/helpers/check_gradient.jl

```diff
@@ -0,0 +1,119 @@
+@doc raw"""
+    check_gradient(M, F, gradF, p=random_point(M), X=random_tangent(M,p))
+
+Check numerivcally whether the gradient `gradF` of `F` is correct.
+
+This implements the method described in Section 4.8 [^Boumal2022].
+
+# Keyword arguments
+
+* `N` (`101`) – number of points to check within the `log_range` default range ``[10^{-8},10^{0}]``
+* `error` - (`false`) throw an error message if the gradient is wrong
+* `io` – (`nothing`) provide an `IO` to print the check result to
+* `limits` (`(10e-8,1)`) specify the limits in the `log_range`
+* `log_range` (`range(limits[1], limits[2]; length=N)`) - specify the range of points (in log scale) to sample the gradient line
+* `plot`- (`false`) whether to plot the resulting check (if `Plots.jl` is loaded)
+* `retraction_method` - (`default_retraction_method(M)`) retraction method to use for the check
+* `slope_tol` – (`0.1`) tolerance for the slope (global) of the approximation
+
+[^Boumal2022]:
+    > Boumal, N.: _An Introduction to Optimization on Smooth Manifolds_, book in preparation,
+    > 2022. url: [http://www.nicolasboumal.net/book](http://www.nicolasboumal.net/book).
+"""
+function check_gradient(
+    M::AbstractManifold,
+    F,
+    gradF,
+    p=random_point(M),
+    X=random_tangent(M, p);
+    plot=false,
+    error=false,
+    io::Union{IO,Nothing}=nothing,
+    limits=(-8.0, 0.0),
+    N=101,
+    log_range=range(limits[1], limits[2]; length=N),
+    retraction_method=default_retraction_method(M),
+    slope_tol=0.1,
+    window=nothing,
+)
+    gradient = gradF(M, p)
+    Xn = X ./ norm(M, p, X) # normalize tangent direction
+    is_vector(M, p, gradient, error)
+    # function for the directional derivative
+    df(M, p, Y) = inner(M, p, gradient, Y)
+    #
+    T = exp10.(log_range)
+    # points p_i to evaluate our error function at
+    points = map(t -> retract(M, p, Xn, t, retraction_method), T)
+    # F(p_i)
+    costs = [F(M, pi) for pi in points]
+    # linearized
+    linearized = map(t -> F(M, p) + t * df(M, p, Xn), T)
+    L = abs.(costs .- linearized)
+    # global fit a + bx
+    x = log_range[L .> 0]
+    y = log10.(L[L .> 0])
+    (a, b) = find_best_slope_window(x, y, length(x))[1:2]
+    if isapprox(b, 2.0; atol=slope_tol)
+        plot &&
+            plot_slope(T[L .> 0], L[L .> 0]; line_base=costs[1], a=a, b=b, i=1, j=length(y))
```

Comment:
`line_base` should be `L[1]`, as otherwise, the dashed line is on a completely different scale.

Reply:
Ah, nice - this puts the dashed line atop the right slope if _all_ values fit, which is a nice feature as well.
I was going for the approach from the Matlab code where the dashed line is somewhere but one can still see the slope in comparison.